File: android/app/src/main/kotlin/com/wagmilabs/sigil/ui/pairing/PairFlowView.kt
package com.wagmilabs.sigil.ui.pairing

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Error
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewmodel.compose.viewModel

@Composable
fun PairFlowView(
    viewModel: PairFlowViewModel = viewModel(),
    modifier: Modifier = Modifier
) {
    val state by viewModel.state.collectAsState()
    val timeRemaining by viewModel.timeRemaining.collectAsState()
    val errorMessage by viewModel.errorMessage.collectAsState()

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Pair Device") }
            )
        }
    ) { padding ->
        Box(
            modifier = modifier
                .fillMaxSize()
                .padding(padding),
            contentAlignment = Alignment.Center
        ) {
            when (val currentState = state) {
                is PairState.Idle -> {
                    ServerURLInputView(
                        errorMessage = errorMessage,
                        onStartPair = { url -> viewModel.startPair(url) }
                    )
                }
                is PairState.Loading -> {
                    LoadingView(message = "Initiating handshake...")
                }
                is PairState.AwaitingConfirmation -> {
                    SessionPictogramView(
                        pictogram = currentState.pictogram,
                        timeRemaining = timeRemaining,
                        onConfirm = { viewModel.confirmPair() },
                        onDeny = { viewModel.denyPair() }
                    )
                }
                is PairState.Completing -> {
                    LoadingView(message = "Completing pair...")
                }
                is PairState.AwaitingAdminApproval -> {
                    AwaitingAdminApprovalView(
                        serverId = currentState.serverId,
                        pollAttempts = currentState.pollAttempts
                    )
                }
                is PairState.Paired -> {
                    PairedView(
                        serverId = currentState.server.serverId,
                        onDone = { viewModel.reset() }
                    )
                }
                is PairState.Denied -> {
                    DeniedView(
                        onTryAgain = { viewModel.reset() }
                    )
                }
                is PairState.Error -> {
                    ErrorView(
                        message = currentState.message,
                        onTryAgain = { viewModel.reset() }
                    )
                }
            }
        }
    }
}

@Composable
private fun ServerURLInputView(
    errorMessage: String?,
    onStartPair: (String) -> Unit,
    modifier: Modifier = Modifier
) {
    var serverUrl by remember { mutableStateOf("") }

    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(24.dp)
            .widthIn(max = 600.dp),
        verticalArrangement = Arrangement.spacedBy(24.dp)
    ) {
        Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
            Text(
                text = "Enter Server URL",
                style = MaterialTheme.typography.titleLarge,
                fontWeight = FontWeight.SemiBold
            )

            Text(
                text = "Enter the Sigil Auth server URL you want to pair with:",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }

        Column(verticalArrangement = Arrangement.spacedBy(16.dp)) {
            OutlinedTextField(
                value = serverUrl,
                onValueChange = { serverUrl = it },
                label = { Text("Server URL") },
                placeholder = { Text("https://sigil.example.com") },
                modifier = Modifier.fillMaxWidth(),
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Uri),
                singleLine = true,
                isError = errorMessage != null
            )

            if (errorMessage != null) {
                Text(
                    text = errorMessage,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.error
                )
            }

            Button(
                onClick = { onStartPair(serverUrl) },
                modifier = Modifier
                    .fillMaxWidth()
                    .heightIn(min = 56.dp),
                enabled = serverUrl.isNotBlank()
            ) {
                Text("Start Pair")
            }
        }
    }
}

@Composable
private fun LoadingView(
    message: String,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier.padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        CircularProgressIndicator()
        Text(
            text = message,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

@Composable
private fun PairedView(
    serverId: String,
    onDone: () -> Unit,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(24.dp)
            .widthIn(max = 600.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(24.dp)
    ) {
        Icon(
            imageVector = Icons.Default.CheckCircle,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.primary,
            modifier = Modifier.size(64.dp)
        )

        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = "Pairing Successful",
                style = MaterialTheme.typography.titleLarge,
                fontWeight = FontWeight.SemiBold
            )

            Text(
                text = "Your device is now paired with $serverId",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center
            )
        }

        Button(
            onClick = onDone,
            modifier = Modifier
                .fillMaxWidth()
                .heightIn(min = 56.dp)
        ) {
            Text("Done")
        }
    }
}

@Composable
private fun DeniedView(
    onTryAgain: () -> Unit,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(24.dp)
            .widthIn(max = 600.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(24.dp)
    ) {
        Icon(
            imageVector = Icons.Default.Warning,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.tertiary,
            modifier = Modifier.size(64.dp)
        )

        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = "Pairing Denied",
                style = MaterialTheme.typography.titleLarge,
                fontWeight = FontWeight.SemiBold
            )

            Text(
                text = "You denied the pairing request or it expired.",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center
            )
        }

        Button(
            onClick = onTryAgain,
            modifier = Modifier
                .fillMaxWidth()
                .heightIn(min = 56.dp)
        ) {
            Text("Try Again")
        }
    }
}

@Composable
private fun ErrorView(
    message: String,
    onTryAgain: () -> Unit,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(24.dp)
            .widthIn(max = 600.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(24.dp)
    ) {
        Icon(
            imageVector = Icons.Default.Error,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.error,
            modifier = Modifier.size(64.dp)
        )

        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = "Error",
                style = MaterialTheme.typography.titleLarge,
                fontWeight = FontWeight.SemiBold
            )

            Text(
                text = message,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center
            )
        }

        Button(
            onClick = onTryAgain,
            modifier = Modifier
                .fillMaxWidth()
                .heightIn(min = 56.dp)
        ) {
            Text("Try Again")
        }
    }
}

File: ios/Sources/UI/Pairing/PairFlowView.swift
import SwiftUI

public struct PairFlowView: View {
    @StateObject private var viewModel: PairFlowViewModel
    @State private var serverURLText: String = ""
    @FocusState private var isServerURLFocused: Bool

    public init(
        networkService: NetworkService,
        pictogramDerivation: SessionPictogramDerivation,
        cryptoService: CryptoService,
        trustStorage: TrustStorageService
    ) {
        _viewModel = StateObject(wrappedValue: PairFlowViewModel(
            networkService: networkService,
            pictogramDerivation: pictogramDerivation,
            cryptoService: cryptoService,
            trustStorage: trustStorage
        ))
    }

    public var body: some View {
        NavigationView {
            ZStack {
                switch viewModel.state {
                case .idle:
                    serverURLInputView
                case .loading:
                    loadingView
                case .awaitingConfirmation(let pictogram, _, _, _, _, _):
                    SessionPictogramView(
                        pictogram: pictogram,
                        timeRemaining: viewModel.timeRemaining,
                        onConfirm: {
                            Task { await viewModel.confirmPair() }
                        },
                        onDeny: {
                            viewModel.denyPair()
                        }
                    )
                case .completing:
                    completingView
                case .awaitingAdminApproval(_, let serverId, _, _, let pollAttempts):
                    AwaitingAdminApprovalView(
                        serverId: serverId,
                        pollAttempts: pollAttempts
                    )
                case .paired(let server):
                    pairedView(server: server)
                case .denied:
                    deniedView
                case .error(let message):
                    errorView(message: message)
                }
            }
            .navigationTitle("Pair Device")
            .navigationBarTitleDisplayMode(.large)
        }
    }

    private var serverURLInputView: some View {
        VStack(spacing: 24) {
            VStack(alignment: .leading, spacing: 8) {
                Text("Enter Server URL")
                    .font(.title2)
                    .fontWeight(.semibold)

                Text("Enter the Sigil Auth server URL you want to pair with:")
                    .font(.body)
                    .foregroundColor(.secondary)
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            VStack(spacing: 16) {
                TextField("https://sigil.example.com", text: $serverURLText)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                    .keyboardType(.URL)
                    .textFieldStyle(.roundedBorder)
                    .focused($isServerURLFocused)

                Button("Start Pair") {
                    startPair()
                }
                .buttonStyle(.borderedProminent)
                .controlSize(.large)
                .disabled(serverURLText.isEmpty)
            }

            if let errorMessage = viewModel.errorMessage {
                Text(errorMessage)
                    .font(.caption)
                    .foregroundColor(.red)
                    .frame(maxWidth: .infinity, alignment: .leading)
            }
        }
        .padding(24)
        .frame(maxWidth: 600)
    }

    private var loadingView: some View {
        VStack(spacing: 16) {
            ProgressView()
            Text("Initiating handshake...")
                .font(.body)
                .foregroundColor(.secondary)
        }
    }

    private var completingView: some View {
        VStack(spacing: 16) {
            ProgressView()
            Text("Completing pair...")
                .font(.body)
                .foregroundColor(.secondary)
        }
    }

    private func pairedView(server: TrustedServer) -> some View {
        VStack(spacing: 24) {
            Image(systemName: "checkmark.circle.fill")
                .font(.system(size: 64))
                .foregroundColor(.green)

            VStack(spacing: 8) {
                Text("Pairing Successful")
                    .font(.title2)
                    .fontWeight(.semibold)

                Text("Your device is now paired with \(server.serverId)")
                    .font(.body)
                    .foregroundColor(.secondary)
                    .multilineTextAlignment(.center)
            }

            Button("Done") {
                viewModel.reset()
            }
            .buttonStyle(.borderedProminent)
            .controlSize(.large)
        }
        .padding(24)
        .frame(maxWidth: 600)
    }

    private var deniedView: some View {
        VStack(spacing: 24) {
            Image(systemName: "xmark.circle.fill")
                .font(.system(size: 64))
                .foregroundColor(.orange)

            VStack(spacing: 8) {
                Text("Pairing Denied")
                    .font(.title2)
                    .fontWeight(.semibold)

                Text("You denied the pairing request or it expired.")
                    .font(.body)
                    .foregroundColor(.secondary)
                    .multilineTextAlignment(.center)
            }

            Button("Try Again") {
                viewModel.reset()
            }
            .buttonStyle(.borderedProminent)
            .controlSize(.large)
        }
        .padding(24)
        .frame(maxWidth: 600)
    }

    private func errorView(message: String) -> some View {
        VStack(spacing: 24) {
            Image(systemName: "exclamationmark.triangle.fill")
                .font(.system(size: 64))
                .foregroundColor(.red)

            VStack(spacing: 8) {
                Text("Error")
                    .font(.title2)
                    .fontWeight(.semibold)

                Text(message)
                    .font(.body)
                    .foregroundColor(.secondary)
                    .multilineTextAlignment(.center)
            }

            Button("Try Again") {
                viewModel.reset()
            }
            .buttonStyle(.borderedProminent)
            .controlSize(.large)
        }
        .padding(24)
        .frame(maxWidth: 600)
    }

    private func startPair() {
        isServerURLFocused = false

        guard var urlString = serverURLText.trimmingCharacters(in: .whitespacesAndNewlines),
              !urlString.isEmpty else {
            return
        }

        if !urlString.hasPrefix("http://") && !urlString.hasPrefix("https://") {
            urlString = "https://" + urlString
        }

        guard let url = URL(string: urlString) else {
            viewModel.errorMessage = "Invalid URL format"
            return
        }

        Task {
            await viewModel.startPair(serverURL: url)
        }
    }
}

#if DEBUG
struct PairFlowView_Previews: PreviewProvider {
    static var previews: some View {
        PairFlowView(
            networkService: DefaultNetworkService(),
            pictogramDerivation: DefaultSessionPictogramDerivation(
                argon2: DefaultArgon2Service(),
                pictogramPool: PictogramPool.shared
            ),
            cryptoService: CryptoServiceImpl(),
            trustStorage: KeychainTrustStorage()
        )
        .previewLayout(.sizeThatFits)
    }
}
#endif
